lib/SIL/SILBasicBlock.cpp

```diff
@@ -187,9 +187,32 @@ transferNodesFromList(llvm::ilist_traits<SILBasicBlock> &SrcTraits,
   // If splicing blocks not in the same function, update the parent pointers.
   for (; First != Last; ++First) {
     First->Parent = Parent;
-    for (auto &II : *First)
+    for (auto &II : *First) {
       II.setDebugScope(B,
                        ScopeCloner.getOrCreateClonedScope(II.getDebugScope()));
+      SILInstruction *I = &II;
+      // Maintain the registrations of opened archetypes.
+      if (isa<OpenExistentialAddrInst>(I) || isa<OpenExistentialRefInst>(I) ||
+          isa<OpenExistentialBoxInst>(I)) {
+        auto Ty = I->getType().getSwiftRValueType();
+        if (isa<ArchetypeType>(Ty) &&
+            !cast<ArchetypeType>(Ty)->getOpenedExistentialType().isNull()) {
+          SrcTraits.Parent->removeOpenedArchetypeDef(Ty, I);
+          Parent->addOpenedArchetypeDef(Ty, I);
+        }
+      } else if (isa<OpenExistentialMetatypeInst>(I)) {
+        SILType InstanceTy = I->getType();
+        while (isa<AnyMetatypeType>(InstanceTy.getSwiftRValueType())) {
+          InstanceTy = InstanceTy.getMetatypeInstanceType(I->getModule());
+        }
+        auto Ty = InstanceTy.getSwiftRValueType();
+        if (isa<ArchetypeType>(Ty) &&
+            !cast<ArchetypeType>(Ty)->getOpenedExistentialType().isNull()) {
+          SrcTraits.Parent->removeOpenedArchetypeDef(Ty, I);
+          Parent->addOpenedArchetypeDef(Ty, I);
+        }
+      }
+    }
```

Comment:
This needs more comments. If we introduce this map, whether inside SILFunction or as a separate analysis, we need  a clear explanation in the code for why it is necessary.

It seems to only be needed when we substitute an opened type in place of a contextual archetype. I think that only happens during inlining (not specialization/devirtualization). However, for inlining, the call-site apply would already have the extra type-use operand, so why would we need a map to find the open_existential? I'm probably just missing something, but so will anyone else reading this code.